doc/design/block.md

```diff
@@ -0,0 +1,236 @@
+a_out = submodel("model_a", a)
+b_out = submodel("model_b", b)
+c_out = submodel("model_c", c)
+
+
+# d is some op's output
+d = someop()
+
+# reuse model_a
+d_model_a_out = submodel("model_a", d)
+```
+
+with `pd.namespace`, user create 3 different namespaces and in each one, the parameters those have specific name will be reused.
+
+## Why use Block to replace NetOp
+- It is weird to treat Network as an operator, but Block in programming language means grouping multiple operators and acts like a single operator, treat a Block as an operator seems more reasonable.
+- The basic concepts including`Variable`, `Operator` and `Block` are concepts in programming language, but `NetOp` is not.
+- Block is a concept in programming language, it behaves similar to current NetOp, but much concrete, and can guide our futher design.
+  - a block looks like something with a pair of curly bracket `{}`, in python we can use `with` statement to make a dramatic code block.
+  - a block will contains its operators(just like NetOp) and some local variables
+  - operator inside a block can make operation on global variables
+  
+## Implementation
+Currentlly, there is a simple implementation in [User Interface Design](), which has the most function of a Block when user writes a program. 
+But it relays on `NetOp` when executes and that works.
+
+We may rename `NetOp` in cpp code to `Block`, but whether to add more functions of a real block into that implementation depends on whether we want to have a `VarDesc` and split the compilation and execution period.
+
+In my opition, Block is very basic concept in underlying implementation, so a python wrapper and a cpp minimal implementation is enough now.
+It is free to change if other module needs more complex supports such as `VarDesc` is added.
+
+If `VerDesc` is added into the framework, Block will be implemented based on NetOp.
+
+**Each Block will have its own Scope, a child-block's scope will inherit father-block's scope.**
+
+
+```c++
+struct VarDesc;
+struct BlockDesc;
+
+// VarDescLib is similar to Scope, but offer a simpler map with the same value type.
+class VarDescLib {
+ public:
+  VarDescLib();
+  // lookup a variable description, recursively or not.
+  VarDesc* Lookup(const std::string& name, bool recursive=true);
+  // create a new description in local namescope.
+  VarDesc* New(const std::string& name);
+  // delete a local variable's description
+  void Delete(const std::string& name);
+ private:
+  std::map<std::string, VarDesc> map_;
+  VarDescLib* father_{nullptr};
+};
+
+/*
+ * Block represents a group of operators.
+ */
+class Block : public OperatorBase {
+ public:
+  // Block which will create its own scope when run
+  Block(Block* father) : father_block_(father) {}
+
+  // Block with a given scope
+  Block(Block* father, Scope* scope) : father_block_(father), scope_(scope) {}
+
+  // NOTE Block has its own scope, so the argument scope here is not needed.
+  virtual void Run(const Scope& scope,
```

Comment:
If Block is an operator, the Run interface will always get a scope, what's the relationship between this scope and Block's own scope that is created or passed in by the constructor?
@wangkuiyi @jacquesqiao